storage-proofs-update/src/vanilla.rs

```diff
@@ -0,0 +1,1209 @@
+            out_data_path_metadata.len()
+        );
+        info!(
+            "replica path {:?}, len {}",
+            replica_path,
+            replica_path_metadata.len()
+        );
+        info!(
+            "sector key path {:?}, len {}",
+            sector_key_path,
+            sector_key_path_metadata.len()
+        );
+
+        // Setup writable mmap for new_replica_path output.
+        let mut out_data = mmap_write(out_data_path)?;
+
+        // Setup read-only mmaps for sector_key_path and staged_data_path inputs.
+        let replica_data = mmap_read(replica_path)?;
+        let sector_key_data = mmap_read(sector_key_path)?;
+
+        let comm_r_old = <TreeRHasher as Hasher>::Function::hash2(&comm_c, &comm_sector_key);
+        let phi = phi(&comm_d_new, &comm_r_old);
+
+        let end = replica_path_metadata.len() as u64;
+
+        // chunk_size is the number of Fr elements to process in parallel chunks.
+        let chunk_size: usize = std::cmp::min(base_tree_nodes_count, CHUNK_SIZE_MIN);
+
+        // data_block_size is the segment length that we're processing
+        // in Fr elements (i.e. chunk_size * sizeof(Fr)).
+        let data_block_size: usize = chunk_size * FR_SIZE;
+
+        // AND-mask which strips the partition-index `k` from a node-index.
+        let node_index_bit_len = nodes_count.trailing_zeros() as usize;
+        let partition_count = partition_count(nodes_count);
+        let partition_bit_len = partition_count.trailing_zeros() as usize;
+        let remove_k_from_node_index_mask = (1 << (node_index_bit_len - partition_bit_len)) - 1;
+
+        Vec::from_iter((0..end).step_by(data_block_size))
+            .into_par_iter()
+            .zip(out_data.par_chunks_mut(data_block_size))
+            .try_for_each(|(chunk_index, output_data)| -> Result<()> {
+                for i in (0..data_block_size as u64).step_by(FR_SIZE) {
+                    let input_index = (chunk_index as usize) + i as usize;
+                    let output_index = i as usize;
+
+                    // Get the `h` high bits from the node-index (sans partition-index).
+                    let node_index = input_index / FR_SIZE;
+                    let input_index_without_k = node_index & remove_k_from_node_index_mask;
+                    let high =
+                        input_index_without_k >> (node_index_bit_len - partition_bit_len - h);
+                    let rho: Fr = <TreeRHasher as Hasher>::Function::hash2(
+                        &phi,
+                        &Fr::from(high as u64).into(),
+                    )
+                    .into();
+
+                    let sector_key_fr =
+                        bytes_into_fr(&sector_key_data[input_index..input_index + FR_SIZE])?;
+                    let replica_data_fr =
+                        bytes_into_fr(&replica_data[input_index..input_index + FR_SIZE])?;
+
+                    let out_data_fr = (replica_data_fr - sector_key_fr) * rho.invert().unwrap();
```

Comment:
Same here, we don't want to compute \rho and invert it on a per-node basis.

Reply:
I believe that I've made the change, but a second review would be beneficial.